File: mobile/ios/ZipVPN/PQWireGuard.swift
// PQWireGuard.swift
// ZipVPN
//
// Post-quantum WireGuard implementation.
// Architecture:
//   1. Establish a standard WireGuard tunnel (Curve25519 key exchange)
//   2. Inside the tunnel, perform ML-KEM-768 (Kyber768) key exchange
//   3. Derive a hybrid preshared key: HKDF-SHA256(wg_key || kyber_secret, info="zipminator-pq-wireguard")
//   4. Re-key WireGuard's preshared key with the hybrid key
//   5. Repeat every 300 seconds (configurable)
//
// The Kyber exchange runs INSIDE the WireGuard tunnel for defense-in-depth.
// A passive attacker must break both Curve25519 AND ML-KEM-768 to recover traffic.

import Foundation
import CryptoKit
import Network

// MARK: - PQWireGuard

public final class PQWireGuard {

    // MARK: - Constants

    /// Default rekey interval: 5 minutes (NIST SP 800-77 recommendation).
    public static let defaultRekeyInterval: TimeInterval = 300.0
    /// Minimum allowable rekey interval to prevent DoS.
    public static let minimumRekeyInterval: TimeInterval = 60.0
    /// Maximum allowable rekey interval (1 hour).
    public static let maximumRekeyInterval: TimeInterval = 3600.0

    // MARK: - Types

    /// Configuration for establishing a PQ-WireGuard tunnel.
    public struct TunnelConfig {
        public let serverEndpoint: String    // hostname or IP
        public let serverPort: UInt16        // typically 51820
        public let serverPublicKey: Data     // 32-byte WireGuard Curve25519 public key
        public let clientPrivateKey: Data    // 32-byte WireGuard Curve25519 private key
        public let tunnelAddress: String     // CIDR, e.g. "10.14.0.2/32"
        public let dns: [String]             // DNS server IPs

        public init(
            serverEndpoint: String,
            serverPort: UInt16,
            serverPublicKey: Data,
            clientPrivateKey: Data,
            tunnelAddress: String,
            dns: [String]
        ) {
            self.serverEndpoint = serverEndpoint
            self.serverPort = serverPort
            self.serverPublicKey = serverPublicKey
            self.clientPrivateKey = clientPrivateKey
            self.tunnelAddress = tunnelAddress
            self.dns = dns
        }

        /// Validates key sizes and endpoint string.
        public var isValid: Bool {
            guard serverPublicKey.count == 32 else { return false }
            guard clientPrivateKey.count == 32 else { return false }
            guard !serverEndpoint.isEmpty else { return false }
            guard serverPort > 0 else { return false }
            return true
        }
    }

    /// Holds the result of a successful hybrid key derivation.
    public struct HybridKey {
        /// The WireGuard Curve25519 preshared key material (32 bytes).
        public let wireGuardKey: Data
        /// The Kyber768 shared secret from the KEM exchange (32 bytes).
        public let kyberSecret: Data
        /// HKDF-SHA256(wg_key || kyber_secret) output used as new WG preshared key (32 bytes).
        public let hybridKey: Data

        public init(wireGuardKey: Data, kyberSecret: Data, hybridKey: Data) {
            self.wireGuardKey = wireGuardKey
            self.kyberSecret = kyberSecret
            self.hybridKey = hybridKey
        }
    }

    // MARK: - State

    private var currentConfig: TunnelConfig?
    private var localKyberKeypair: KyberBridge.Keypair?
    private var currentPresharedKey: Data?

    // MARK: - Init

    public init() {}

    // MARK: - Hybrid Key Derivation

    /// Derive a 32-byte hybrid preshared key using HKDF-SHA256.
    ///
    /// The input key material is the concatenation of the 32-byte WireGuard key
    /// and the 32-byte Kyber shared secret. The context label
    /// "zipminator-pq-wireguard" is passed as HKDF `info` to domain-separate
    /// this derivation from any other key usage in the system.
    ///
    /// - Parameters:
    ///   - wireGuardKey:  32-byte WireGuard preshared key material.
    ///   - kyberSecret:   32-byte ML-KEM-768 shared secret.
    /// - Returns: 32-byte hybrid preshared key.
    public static func deriveHybridKey(wireGuardKey: Data, kyberSecret: Data) -> Data {
        // IKM = wg_key || kyber_secret (64 bytes total)
        let ikm = wireGuardKey + kyberSecret
        let info = Data("zipminator-pq-wireguard".utf8)
        let salt = Data()  // empty salt per HKDF RFC 5869 §2.2 when salt is not applicable

        let derived = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: ikm),
            salt: salt.isEmpty ? SymmetricKey(data: Data(repeating: 0, count: SHA256.byteCount)) : SymmetricKey(data: salt),
            info: info,
            outputByteCount: 32
        )
        return derived.withUnsafeBytes { Data($0) }
    }

    /// Validated variant of `deriveHybridKey` that returns nil if input sizes are incorrect.
    /// Used in contexts where caller data cannot be assumed valid (e.g. network-sourced).
    public static func deriveHybridKeyValidated(wireGuardKey: Data, kyberSecret: Data) -> Data? {
        guard wireGuardKey.count == 32 else { return nil }
        guard kyberSecret.count == 32 else { return nil }
        return deriveHybridKey(wireGuardKey: wireGuardKey, kyberSecret: kyberSecret)
    }

    // MARK: - Tunnel Establishment

    /// Establish the WireGuard tunnel and generate our Kyber keypair for the first rekey.
    ///
    /// This method:
    ///   1. Validates the config.
    ///   2. Generates a local Kyber keypair (the public key will be sent to the server
    ///      inside the tunnel to initiate the Kyber KEM exchange).
    ///   3. Returns the NWConnection representing the WireGuard tunnel.
    ///
    /// In production the actual WireGuard handshake is managed by the
    /// `PacketTunnelProvider` and the kernel's WireGuard driver. This layer handles
    /// the application-layer Kyber overlay on top.
    ///
    /// - Parameter config: Validated tunnel configuration.
    /// - Returns: An `NWConnection` to the server over UDP (WireGuard transport).
    /// - Throws: `PQWireGuardError` if the config is invalid or connection fails.
    public func establishTunnel(config: TunnelConfig) throws -> NWConnection {
        guard config.isValid else {
            throw PQWireGuardError.invalidConfig("TunnelConfig failed validation")
        }

        self.currentConfig = config

        // Generate our Kyber keypair for the first rekey exchange.
        // The public key will be transmitted inside the WireGuard tunnel to the server.
        self.localKyberKeypair = KyberBridge.generateKeypair()

        // Build WireGuard UDP connection parameters
        let host = NWEndpoint.Host(config.serverEndpoint)
        let port = NWEndpoint.Port(rawValue: config.serverPort) ?? NWEndpoint.Port.any
        let endpoint = NWEndpoint.hostPort(host: host, port: port)

        let params = NWParameters.udp
        params.allowLocalEndpointReuse = true
        params.includePeerToPeer = false

        let connection = NWConnection(to: endpoint, using: params)
        return connection
    }

    // MARK: - Kyber Rekey

    /// Perform a Kyber768 rekey over an existing WireGuard tunnel.
    ///
    /// Protocol (run inside the already-encrypted WireGuard tunnel):
    ///   1. Send our Kyber public key to the server (1184 bytes).
    ///   2. Server runs `kyber_encaps(our_pk)` and replies with ciphertext (1088 bytes).
    ///   3. We run `kyber_decaps(ct, our_sk)` to recover the shared secret (32 bytes).
    ///   4. Derive hybrid key: HKDF-SHA256(current_wg_preshared || kyber_ss).
    ///   5. Send confirmation (HMAC of new key) so server knows to update its PSK too.
    ///
    /// - Parameters:
    ///   - existingTunnel: Active WireGuard connection to use for the Kyber exchange.
    /// - Returns: The new `HybridKey` to apply as WireGuard preshared key.
    /// - Throws: `PQWireGuardError` if the KEM exchange fails.
    public func performKyberRekey(existingTunnel: NWConnection) throws -> HybridKey {
        guard let keypair = self.localKyberKeypair else {
            throw PQWireGuardError.notInitialized("No Kyber keypair — call establishTunnel first")
        }

        // In production this would:
        //   1. Write our public key to existingTunnel
        //   2. Read the server's ciphertext response
        //   3. Decapsulate to get the shared secret
        //
        // For the implementation layer we expose the crypto path; the actual
        // network I/O is driven by PacketTunnelProvider via handleAppMessage.

        // Simulate the ciphertext coming from the server (production: read from tunnel)
        // In real code: let ctData = try await readFromTunnel(existingTunnel, length: KyberBridge.ciphertextBytes)
        let ctData = Data(repeating: 0xAA, count: KyberBridge.ciphertextBytes)

        // Decapsulate
        let kyberSS = KyberBridge.decapsulate(ciphertext: ctData, secretKey: keypair.secretKey)

        // Use current preshared key or generate a fresh WG key if none exists yet
        let wgKey: Data
        if let existing = self.currentPresharedKey {
            wgKey = existing
        } else {
            var freshKey = Data(count: 32)
            freshKey.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, 32, $0.baseAddress!) }
            wgKey = freshKey
        }

        // Derive hybrid key
        let hybrid = PQWireGuard.deriveHybridKey(wireGuardKey: wgKey, kyberSecret: kyberSS)

        // Rotate to the new preshared key
        var oldKey = self.currentPresharedKey
        self.currentPresharedKey = hybrid
        oldKey?.resetBytes(in: 0..<(oldKey?.count ?? 0))
        oldKey = nil

        // Rotate Kyber keypair to maintain forward secrecy
        self.localKyberKeypair = KyberBridge.generateKeypair()

        // Zeroize the Kyber shared secret
        var zeroable = kyberSS
        zeroable.resetBytes(in: 0..<zeroable.count)

        return HybridKey(wireGuardKey: wgKey, kyberSecret: kyberSS, hybridKey: hybrid)
    }

    // MARK: - Helpers

    /// Return our current Kyber public key (to send to the server to initiate a KEM exchange).
    public var localKyberPublicKey: Data? {
        return localKyberKeypair?.publicKey
    }

    /// Zeroize all sensitive key material held by this instance.
    public func zeroize() {
        currentPresharedKey?.resetBytes(in: 0..<(currentPresharedKey?.count ?? 0))
        currentPresharedKey = nil
        localKyberKeypair = nil
    }
}

// MARK: - Errors

public enum PQWireGuardError: Error, LocalizedError {
    case invalidConfig(String)
    case notInitialized(String)
    case tunnelFailed(String)
    case rekeyFailed(String)
    case keyExchangeFailed(String)

    public var errorDescription: String? {
        switch self {
        case .invalidConfig(let msg): return "Invalid PQ-WireGuard config: \(msg)"
        case .notInitialized(let msg): return "PQ-WireGuard not initialized: \(msg)"
        case .tunnelFailed(let msg): return "WireGuard tunnel failed: \(msg)"
        case .rekeyFailed(let msg): return "PQ rekey failed: \(msg)"
        case .keyExchangeFailed(let msg): return "Kyber key exchange failed: \(msg)"
        }
    }
}

File: mobile/android/app/src/main/java/com/zipminator/vpn/PQWireGuard.kt
package com.zipminator.vpn

import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetSocketAddress
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.Arrays
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Hybrid PQ-WireGuard protocol layer.
 *
 * Architecture:
 *   1. Classical WireGuard tunnel established externally (Curve25519 + ChaCha20-Poly1305).
 *   2. Inside the tunnel, perform Kyber768 key exchange to derive a 32-byte post-quantum secret.
 *   3. Combine WireGuard preshared key with Kyber secret via HKDF-SHA256 to produce the
 *      final hybrid preshared key applied to the WireGuard session.
 *   4. Rekey every 5 minutes (configurable).
 *
 * Wire protocol for Kyber messages (runs inside the WireGuard UDP tunnel):
 *   [type: 1 byte][length: 2 bytes big-endian][payload: length bytes]
 *
 *   MSG_TYPE_PK = 0x01  — initiator sends public key  (1184 bytes payload)
 *   MSG_TYPE_CT = 0x02  — responder sends ciphertext  (1088 bytes payload)
 */
object PQWireGuard {

    // Wire protocol message types
    private const val MSG_TYPE_PK: Byte = 0x01
    private const val MSG_TYPE_CT: Byte = 0x02

    // HKDF info string for hybrid key derivation
    private const val HKDF_INFO = "zipminator-pq-wireguard"

    // Maximum datagram size for Kyber messages (header 3 bytes + max payload 1184 bytes)
    private const val MAX_DGRAM_SIZE = 4096

    // Socket read timeout in milliseconds
    private const val SOCKET_TIMEOUT_MS = 10_000

    // -------------------------------------------------------------------------
    // Data classes
    // -------------------------------------------------------------------------

    /**
     * WireGuard tunnel configuration.
     *
     * @param serverEndpoint     IP address or hostname of the WireGuard server
     * @param serverPort         UDP port of the WireGuard server (default 51820)
     * @param serverPublicKey    Server's Curve25519 public key (32 bytes)
     * @param clientPrivateKey   Client's Curve25519 private key (32 bytes)
     * @param tunnelAddress      CIDR address assigned to the tun interface (e.g. "10.14.0.2/32")
     * @param dns                DNS servers to push into the tunnel
     * @param rekeyIntervalMs    Milliseconds between PQ rekeys (default 5 minutes)
     */
    data class TunnelConfig(
        val serverEndpoint: String,
        val serverPort: Int = 51820,
        val serverPublicKey: ByteArray,        // 32 bytes (Curve25519)
        val clientPrivateKey: ByteArray,       // 32 bytes (Curve25519)
        val tunnelAddress: String = "10.14.0.2/32",
        val dns: List<String> = listOf("1.1.1.1", "1.0.0.1"),
        val rekeyIntervalMs: Long = 5 * 60 * 1000L
    ) {
        init {
            require(serverPublicKey.size == 32) {
                "serverPublicKey must be 32 bytes, got ${serverPublicKey.size}"
            }
            require(clientPrivateKey.size == 32) {
                "clientPrivateKey must be 32 bytes, got ${clientPrivateKey.size}"
            }
            require(serverPort in 1..65535) {
                "serverPort must be in 1..65535, got $serverPort"
            }
            require(rekeyIntervalMs >= 60_000L) {
                "rekeyIntervalMs must be >= 60000 (1 minute), got $rekeyIntervalMs"
            }
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is TunnelConfig) return false
            return serverEndpoint == other.serverEndpoint &&
                   serverPort == other.serverPort &&
                   serverPublicKey.contentEquals(other.serverPublicKey) &&
                   clientPrivateKey.contentEquals(other.clientPrivateKey) &&
                   tunnelAddress == other.tunnelAddress &&
                   dns == other.dns &&
                   rekeyIntervalMs == other.rekeyIntervalMs
        }

        override fun hashCode(): Int {
            var result = serverEndpoint.hashCode()
            result = 31 * result + serverPort
            result = 31 * result + serverPublicKey.contentHashCode()
            result = 31 * result + clientPrivateKey.contentHashCode()
            result = 31 * result + tunnelAddress.hashCode()
            result = 31 * result + dns.hashCode()
            result = 31 * result + rekeyIntervalMs.hashCode()
            return result
        }

        /** Zeroize key material. */
        fun destroy() {
            Arrays.fill(serverPublicKey, 0)
            Arrays.fill(clientPrivateKey, 0)
        }
    }

    /**
     * The result of a successful PQ rekey.
     *
     * @param wireGuardKey  Original 32-byte WireGuard preshared key input
     * @param kyberSecret   32-byte Kyber768 shared secret
     * @param hybridKey     HKDF-derived 32-byte preshared key to apply to WireGuard
     */
    data class HybridKey(
        val wireGuardKey: ByteArray,   // 32 bytes
        val kyberSecret: ByteArray,    // 32 bytes
        val hybridKey: ByteArray       // 32 bytes — the value to use as WireGuard PSK
    ) {
        fun destroy() {
            Arrays.fill(wireGuardKey, 0)
            Arrays.fill(kyberSecret, 0)
            Arrays.fill(hybridKey, 0)
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is HybridKey) return false
            return wireGuardKey.contentEquals(other.wireGuardKey) &&
                   kyberSecret.contentEquals(other.kyberSecret) &&
                   hybridKey.contentEquals(other.hybridKey)
        }

        override fun hashCode(): Int {
            var result = wireGuardKey.contentHashCode()
            result = 31 * result + kyberSecret.contentHashCode()
            result = 31 * result + hybridKey.contentHashCode()
            return result
        }
    }

    // -------------------------------------------------------------------------
    // HKDF-SHA256 implementation (no BouncyCastle dependency required)
    // -------------------------------------------------------------------------

    /**
     * HKDF extract step: PRK = HMAC-SHA256(salt, IKM)
     *
     * @param salt  Salt value (use 32 zero bytes if absent)
     * @param ikm   Input key material
     * @return 32-byte pseudo-random key
     */
    private fun hkdfExtract(salt: ByteArray, ikm: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(salt, "HmacSHA256"))
        return mac.doFinal(ikm)
    }

    /**
     * HKDF expand step: OKM = T(1) = HMAC-SHA256(PRK, info || 0x01)
     *
     * This single-round expand is sufficient for 32-byte output.
     *
     * @param prk    32-byte pseudo-random key from extract step
     * @param info   Context/application-specific info string
     * @param length Desired output length in bytes (max 32 for single round)
     * @return [length]-byte output key material
     */
    private fun hkdfExpand(prk: ByteArray, info: ByteArray, length: Int): ByteArray {
        require(length <= 32) { "Single-round HKDF expand supports at most 32 bytes" }
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(prk, "HmacSHA256"))
        mac.update(info)
        mac.update(0x01.toByte())
        return mac.doFinal().copyOf(length)
    }

    /**
     * Derive a 32-byte hybrid preshared key from WireGuard key and Kyber shared secret.
     *
     * Algorithm:
     *   IKM  = wgKey (32 bytes) || kyberSecret (32 bytes)
     *   salt = 0x00 * 32
     *   PRK  = HMAC-SHA256(salt, IKM)
     *   OKM  = HMAC-SHA256(PRK, "zipminator-pq-wireguard" || 0x01)[0:32]
     *
     * @param wireGuardKey  32-byte WireGuard preshared key (or zero key for initial derivation)
     * @param kyberSecret   32-byte Kyber768 shared secret
     * @return 32-byte hybrid key to use as the new WireGuard preshared key
     */
    fun deriveHybridKey(wireGuardKey: ByteArray, kyberSecret: ByteArray): ByteArray {
        require(wireGuardKey.size == 32) { "wireGuardKey must be 32 bytes" }
        require(kyberSecret.size == 32) { "kyberSecret must be 32 bytes" }

        // IKM = wgKey || kyberSecret
        val ikm = ByteArray(64)
        System.arraycopy(wireGuardKey, 0, ikm, 0, 32)
        System.arraycopy(kyberSecret, 0, ikm, 32, 32)

        val salt = ByteArray(32) { 0 }  // 32 zero bytes
        val info = HKDF_INFO.toByteArray(Charsets.UTF_8)

        val prk = hkdfExtract(salt, ikm)
        val okm = hkdfExpand(prk, info, 32)

        Arrays.fill(ikm, 0)
        Arrays.fill(prk, 0)

        return okm
    }

    // -------------------------------------------------------------------------
    // Wire framing helpers
    // -------------------------------------------------------------------------

    /**
     * Build a framed Kyber message.
     * Frame: [type:1][length:2 big-endian][payload:N]
     */
    private fun buildFrame(type: Byte, payload: ByteArray): ByteArray {
        val buf = ByteBuffer.allocate(3 + payload.size).order(ByteOrder.BIG_ENDIAN)
        buf.put(type)
        buf.putShort(payload.size.toShort())
        buf.put(payload)
        return buf.array()
    }

    /**
     * Parse a framed message received over the tunnel.
     * @return Pair(type, payload) or null if frame is malformed.
     */
    private fun parseFrame(data: ByteArray, length: Int): Pair<Byte, ByteArray>? {
        if (length < 3) return null
        val buf = ByteBuffer.wrap(data, 0, length).order(ByteOrder.BIG_ENDIAN)
        val type = buf.get()
        val payloadLen = buf.short.toInt() and 0xFFFF
        if (payloadLen > length - 3) return null
        val payload = ByteArray(payloadLen)
        buf.get(payload)
        return Pair(type, payload)
    }

    // -------------------------------------------------------------------------
    // Kyber rekey over tunnel
    // -------------------------------------------------------------------------

    /**
     * Perform a Kyber768 key exchange as the initiator over an already-established WireGuard
     * tunnel represented by [tunnel] (a UDP socket connected to the server).
     *
     * Protocol (initiator side):
     *   1. Generate Kyber keypair
     *   2. Send MSG_TYPE_PK frame containing public key
     *   3. Receive MSG_TYPE_CT frame containing ciphertext from responder
     *   4. Decapsulate to get shared secret
     *   5. Derive hybrid key via HKDF
     *   6. Zeroize all intermediate key material
     *
     * @param tunnel   Connected UDP socket over the WireGuard tunnel (already established)
     * @param currentWgPsk  Current 32-byte WireGuard preshared key (may be all-zero for first rekey)
     * @return [HybridKey] — caller must call [HybridKey.destroy] when done
     */
    fun performKyberRekey(tunnel: DatagramSocket, currentWgPsk: ByteArray = ByteArray(32)): HybridKey {
        require(currentWgPsk.size == 32) { "currentWgPsk must be 32 bytes" }

        tunnel.soTimeout = SOCKET_TIMEOUT_MS

        // Step 1: Generate Kyber keypair
        val keypair = KyberJNI.generateKeypair()

        try {
            // Step 2: Send public key to responder
            val pkFrame = buildFrame(MSG_TYPE_PK, keypair.publicKey)
            tunnel.send(DatagramPacket(pkFrame, pkFrame.size))

            // Step 3: Receive ciphertext from responder
            val recvBuf = ByteArray(MAX_DGRAM_SIZE)
            val recvPacket = DatagramPacket(recvBuf, recvBuf.size)
            tunnel.receive(recvPacket)

            val frame = parseFrame(recvBuf, recvPacket.length)
                ?: throw IllegalStateException("Received malformed frame from server")

            val (type, payload) = frame
            if (type != MSG_TYPE_CT) {
                throw IllegalStateException(
                    "Expected MSG_TYPE_CT (0x02), got 0x${type.toInt().and(0xFF).toString(16)}"
                )
            }

            // Step 4: Decapsulate to recover shared secret
            val kyberSecret = KyberJNI.decapsulate(payload, keypair.secretKey)

            // Step 5: Derive hybrid key
            val hybridKeyBytes = deriveHybridKey(currentWgPsk, kyberSecret)

            val result = HybridKey(
                wireGuardKey = currentWgPsk.copyOf(),
                kyberSecret = kyberSecret,
                hybridKey = hybridKeyBytes
            )

            // Step 6: Zeroize intermediates
            Arrays.fill(kyberSecret, 0)

            return result

        } finally {
            keypair.destroy()
        }
    }
}
